High CPU usage may occur due to infinite loop in WebProcessPool::processForNavigation()

cdumez opened the pull request:
#### 5ea58380f09490bd9a91cc1ab713e5c7a067e4d2
<pre>
High CPU usage may occur due to infinite loop in WebProcessPool::processForNavigation()
<a href="https://bugs.webkit.org/show_bug.cgi?id=266389">https://bugs.webkit.org/show_bug.cgi?id=266389</a>
<a href="https://rdar.apple.com/118280525">rdar://118280525</a>

Reviewed by Alex Christensen.

From the profile in the radar, we can see that we can end up in a state where:
1. WebProcessPool::processForNavigation() send the AddAllowedFirstPartyForCookies
   IPC to the network process
2. Once we received the IPC response from the network process, we check if the
   destination WebProcess is still running.
3. We notice that the WebProcess is no longer running so we call
   WebProcessPool::processForNavigation() -&gt; Back to step 1.

We end up in an infinite loop (with an AddAllowedFirstPartyForCookies async IPC
in the middle) because we determine over and over that the destination process
for the navigation is no longer running, likely because processForNavigation()
keeps selecting the same WebProcess.

To make sure this can no longer happen, I made the following changes:
1. Add a cache to NetworkProcessProxy to avoid sending the
   AddAllowedFirstPartyForCookies async IPC unnecessarily. It is common for
   navigations to happen within the same process when the registrable domain
   doesn&apos;t change. In this case, there is no point in sending the IPC again.
2. Make sure processForNavigationInternal() never returns a terminated process
   by adding a couple of missing checks.
3. Move the AddAllowedFirstPartyForCookies handshake to a new function named
   prepareProcessForNavigation(). This function calls itself recursively when
   the destination process is no longer running, like we used to. However,
   I now make sure to fall back to calling prepareProcessForNavigation() with
   a fresh new WebProcess. Also, we attempt this fall back 3 times at most to
   guarantee no infinite looping can occur.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::addAllowedFirstPartyForCookies):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:

* Source/WebKit/UIProcess/WebNavigationState.cpp:
(WebKit::WebNavigationState::navigation):
Drop assertion that required the navigationID to be present in the map when
calling navigation(navigationID). I changed timing by avoiding the async IPC
handshake with the network process in some cases. This timing change caused
the ProcessSwap.ConcurrentHistoryNavigations API test to hit this assertion
in didStartProvisionalLoad(). The API test purposefully starts parallel /
competing navigations. We used to get lucky and the navigation would get
cancelled before receiving the response from the network process. Now the
cancellation happens after the navigation has started in the provisional
process.

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
(WebKit::WebProcessPool::prepareProcessForNavigation):
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessPool.h:

Canonical link: <a href="https://commits.webkit.org/272059@main">https://commits.webkit.org/272059@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c514c9cdb7d829b03701f0e5499c44416ff8685

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30476 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9148 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32096 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32985 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27566 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11386 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6390 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27510 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30784 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7694 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27318 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6590 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6744 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27165 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34324 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27764 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27662 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32916 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6729 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4872 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30741 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8474 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7464 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3950 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7271 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
```diff
@@ -31,6 +31,7 @@
 #include "DataReference.h"
 #include "DataTaskIdentifier.h"
 #include "IdentifierTypes.h"
+#include "LoadedWebArchive.h"
```
We might be able to just forward declare LoadedWebArchive instead of including this here

cdumez committed:
High CPU usage may occur due to infinite loop in WebProcessPool::processForNavigation()
https://bugs.webkit.org/show_bug.cgi?id=266389
rdar://118280525

Reviewed by Alex Christensen.

From the profile in the radar, we can see that we can end up in a state where:
1. WebProcessPool::processForNavigation() send the AddAllowedFirstPartyForCookies
   IPC to the network process
2. Once we received the IPC response from the network process, we check if the
   destination WebProcess is still running.
3. We notice that the WebProcess is no longer running so we call
   WebProcessPool::processForNavigation() -> Back to step 1.

We end up in an infinite loop (with an AddAllowedFirstPartyForCookies async IPC
in the middle) because we determine over and over that the destination process
for the navigation is no longer running, likely because processForNavigation()
keeps selecting the same WebProcess.

To make sure this can no longer happen, I made the following changes:
1. Add a cache to NetworkProcessProxy to avoid sending the
   AddAllowedFirstPartyForCookies async IPC unnecessarily. It is common for
   navigations to happen within the same process when the registrable domain
   doesn't change. In this case, there is no point in sending the IPC again.
2. Make sure processForNavigationInternal() never returns a terminated process
   by adding a couple of missing checks.
3. Move the AddAllowedFirstPartyForCookies handshake to a new function named
   prepareProcessForNavigation(). This function calls itself recursively when
   the destination process is no longer running, like we used to. However,
   I now make sure to fall back to calling prepareProcessForNavigation() with
   a fresh new WebProcess. Also, we attempt this fall back 3 times at most to
   guarantee no infinite looping can occur.

* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::addAllowedFirstPartyForCookies):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:

* Source/WebKit/UIProcess/WebNavigationState.cpp:
(WebKit::WebNavigationState::navigation):
Drop assertion that required the navigationID to be present in the map when
calling navigation(navigationID). I changed timing by avoiding the async IPC
handshake with the network process in some cases. This timing change caused
the ProcessSwap.ConcurrentHistoryNavigations API test to hit this assertion
in didStartProvisionalLoad(). The API test purposefully starts parallel /
competing navigations. We used to get lucky and the navigation would get
cancelled before receiving the response from the network process. Now the
cancellation happens after the navigation has started in the provisional
process.

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):
(WebKit::WebProcessPool::prepareProcessForNavigation):
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessPool.h:

Canonical link: https://commits.webkit.org/272059@main
